benchmark/common.js

```diff
@@ -1,205 +1,207 @@
+  const self = this;
 
-  args = args.concat(url);
+  const urlFull = 'http://127.0.0.1:' + exports.PORT + urlPath;
+  args = args.concat(urlFull);
 
-  var out = '';
-  var child = child_process.spawn('wrk', args);
+  const childStart = process.hrtime();
+  const child = child_process.spawn('wrk', args);
+  child.stderr.pipe(process.stderr);
 
+  // Collect stdout
+  let stdout = '';
   child.stdout.setEncoding('utf8');
+  child.stdout.on('data', (chunk) => stdout += chunk);
 
-  child.stdout.on('data', function(chunk) {
-    out += chunk;
-  });
-
-  child.on('close', function(code) {
-    if (cb)
-      cb(code);
+  child.once('close', function(code) {
+    const elapsed = process.hrtime(childStart);
+    if (cb) cb(code);
 
     if (code) {
       console.error('wrk failed with ' + code);
       process.exit(code);
+      return;
     }
-    var match = out.match(regexp);
-    var qps = match && +match[1];
-    if (!qps) {
-      console.error('%j', out);
-      console.error('wrk produced strange output');
+
+    // Extract requests pr second and check for odd results
+    const match = stdout.match(WRK_REGEXP);
+    if (!match || match.length <= 1) {
+      console.error('wrk produced strange output:');
+      console.error(stdout);
       process.exit(1);
+      return;
     }
-    self.report(+qps);
+
+    // Report rate
+    self.report(parseFloat(match[1]), elapsed);
   });
 };
 
 Benchmark.prototype._run = function() {
-  if (this.config)
-    return this.fn(this.config);
-
-  // some options weren't set.
-  // run with all combinations
-  var main = require.main.filename;
-  var options = this.options;
-
-  var queue = Object.keys(options).reduce(function(set, key) {
-    var vals = options[key];
-    assert(Array.isArray(vals));
-
-    // match each item in the set with each item in the list
-    var newSet = new Array(set.length * vals.length);
-    var j = 0;
-    set.forEach(function(s) {
-      vals.forEach(function(val) {
-        if (typeof val !== 'number' && typeof val !== 'string') {
-          throw new TypeError(`configuration "${key}" had type ${typeof val}`);
-        }
-
-        newSet[j++] = s.concat(key + '=' + val);
-      });
-    });
-    return newSet;
-  }, [[main]]);
-
-  // output csv heading
-  if (outputFormat == 'csv')
-    console.log('filename,' + Object.keys(options).join(',') + ',result');
-
-  var node = process.execPath;
-  var i = 0;
-  function run() {
-    var argv = queue[i++];
-    if (!argv)
-      return;
-    argv = process.execArgv.concat(argv);
-    var child = child_process.spawn(node, argv, { stdio: 'inherit' });
-    child.on('close', function(code, signal) {
-      if (code)
-        console.error('child process exited with code ' + code);
-      else
-        run();
-    });
-  }
-  run();
-};
+  const self = this;
+
+  // If there is only one configuarion don't spawn a process, just run
+  // the benchmark function
+  if (this.queue.length === 1) return this.fn(this.queue[0]);
```

Comment:
i feel like we should always spawn a process in order to keep measurements uniform as possible. just my opinion though.